apps/dash-study-browser/app.py

```diff
@@ -0,0 +1,322 @@
+                                            ),
+                                        ),
+                                    ],
+                                ),
+                            ],
+                        )
+                    ],
+                ),
+                # Graph
+                html.Div(
+                    className="eight columns card-left",
+                    children=[
+                        html.Div(
+                            className="bg-white",
+                            children=[
+                                html.H5("Animal data plot"),
+                                dcc.Graph(id="plot"),
+                            ],
+                        )
+                    ],
+                ),
+            ],
+        ),
+    ]
+)
+
+# Callback to create error message
+@app.callback(Output("error-message", "children"), [Input("upload-data", "contents")])
+def update_error(contents):
+    if contents:
+        try:
+            content_string = contents.split(",")
+            decoded = base64.b64decode(content_string)
+            study_data = pd.read_csv(io.StringIO(decoded.decode("utf-8")))
+        except Exception as e:
+            print(e)
+            return html.Div(
+                className="alert",
+                children=["That doesn't seem to be a valid csv file!"],
+            )
+    else:
+        study_data = default_study_data
+
+    missing_columns = {
+        "group_id",
+        "group_type",
+        "reading_value",
+        "study_id",
+    }.difference(study_data.columns)
+
+    if missing_columns:
+        return html.Div(
+            className="alert", children=["Missing columns: " + str(missing_columns)]
+        )
+
+    return None
+
+
+# Callback to generate study-dropdown options and values
+@app.callback(
+    [Output("study-dropdown", "options"), Output("study-dropdown", "value")],
+    [Input("error-message", "children")],
+    [State("upload-data", "contents")],
+)
+def update_dropdown(error_message, contents):
+    if error_message:
+        study_data = default_study_data
+    elif contents:
+        content_string = contents.split(",")
+        decoded = base64.b64decode(content_string)
+        study_data = pd.read_csv(io.StringIO(decoded.decode("utf-8")))
+    else:
+        study_data = default_study_data
+
+    options = []
+    if "test_article" in study_data.columns:
+        test_articles = study_data.test_article.unique()
+        for test_article in test_articles:
+            studies = study_data.study_id[
+                study_data.test_article == test_article
+            ].unique()
+            for study in studies:
+                options.append(
+                    {
+                        "label": "{} (study: {})".format(test_article, study),
+                        "value": study,
+                    }
+                )
+    else:
+        studies = study_data.study_id.unique()
+        for study in studies:
```

Comment:
This is preference but if ur using the variable only once.

I prefer `for study in study_data.study_id.unique():`